tests/lib/rook/upstream.py

```diff
@@ -28,81 +29,88 @@ class RookCluster(RookBase):
+                    "make -j BUILD_REGISTRY='rook-build' IMAGES='ceph'"))
+
+        image = 'rook/ceph'
+        tag = f"{settings.UPSTREAM_ROOK.VERSION}-rookcheck"
+        self.rook_image = f"{image}:{tag}"
+        logger.info(f"Tag image as {image}:{tag}")
+        execute(f'docker tag "rook-build/ceph-amd64" {image}:{tag}')
+
+        logger.info("Save image tar")
+        # TODO(jhesketh): build arch may differ
+        execute(f"docker save {image}:{tag} | gzip > %s"
+                % os.path.join(self.build_dir, 'rook-ceph.tar.gz'))
+        self._rook_built = True
 
-            logger.info("[build_rook] Save image tar")
-            # TODO(jhesketh): build arch may differ
-            execute(
-                "docker save rook/ceph:master | gzip > %s"
-                % os.path.join(self.builddir, 'rook-ceph.tar.gz')
-            )
+    def upload_rook_image(self):
+        self.kubernetes.hardware.ansible_run_playbook(
+            "playbook_rook_upstream.yaml")
 
-        self.ceph_dir = os.path.join(
-            self.builddir,
-            'src/github.com/rook/rook/cluster/examples/kubernetes/ceph'
+    def get_rook(self):
+        logger.info("Clone rook repo %s" % settings.UPSTREAM_ROOK.REPO)
+        execute(
+            "git clone %s %s" % (settings.UPSTREAM_ROOK.REPO, self.build_dir),
+            log_stderr=False
+        )
+        os.chdir(self.build_dir)
+        logger.info("Checkout rook version %s"
+                    % settings.UPSTREAM_ROOK.VERSION)
+        execute("git checkout %s" % settings.UPSTREAM_ROOK.VERSION,
+                log_stderr=False)
+
+    def get_golang(self):
+        url = 'https://golang.org/VERSION?m=text'
+        version = requests.get(url).content.decode("utf-8")
+        self.workspace.get_unpack(
+            "https://dl.google.com/go/%s.linux-amd64.tar.gz" % version,
+            unpack_folder=self.workspace.bin_dir
         )
 
-        self._rook_built = True
-
-    def preinstall(self):
-        super().preinstall()
-        if converter('@bool', settings.UPSTREAM_ROOK.BUILD_ROOK_FROM_GIT):
-            self.upload_rook_image()
-
-    def upload_rook_image(self):
-        self.kubernetes.hardware.ansible_run_playbook("playbook_rook.yaml")
+    def get_helm(self):
+        url = "https://api.github.com/repos/helm/helm/releases/latest"
+        version = requests.get(url).json()["tag_name"]
+        self.workspace.get_unpack(
+            "https://get.helm.sh/helm-%s-linux-amd64.tar.gz" % version)
+        os.rename(os.path.join(self.workspace.tmp_dir, 'linux-amd64', 'helm'),
+                  os.path.join(self.workspace.bin_dir, 'helm'))
+
+    def _fix_yaml(self):
+        # Replace image reference if we built it in this run
+        with open(os.path.join(self.ceph_dir, 'operator.yaml')) as file:
+            docs = yaml.load_all(file, Loader=yaml.FullLoader)
+            for doc in docs:
+                try:
```

Comment:
(nit) I would loop until `doc['kind'] == 'Deployment' and doc['metadata']['name'] == 'rook-ceph-operator'`, then lookup the image directly (as `doc['spec']['template']['spec']['containers'][0]['image']`) without catching the KeyError so that we fail the build if we don't do this replacement correctly.

Reply:
The KeyError catch is because there are two documents in one yaml fil actually.
So the key is not present in both.
If the key is not present in any it will fail witch is what we want. 
Is that OK ?

Reply:
Right, so my suggestion was to find the right document by matching on the `kind` and `metadata.name`. Once you have the right document we assume the key has to be there.

But yes, your implementation will work as well, I was just being nit-picky ;-)